File: sdk/cpp/examples/basic_example.cpp
#include "kvdb_client.h"
#include <iostream>
#include <thread>
#include <chrono>

using namespace kvdb::client;

int main() {
    // 配置客户端
    ClientConfig config;
    config.server_address = "localhost:50051";
    config.protocol = "grpc";
    config.connection_timeout_ms = 5000;
    config.request_timeout_ms = 10000;
    
    // 创建客户端
    auto client = ClientFactory::create_client(config);
    
    // 连接到服务器
    auto connect_result = client->connect();
    if (!connect_result.success) {
        std::cerr << "Failed to connect: " << connect_result.error_message << std::endl;
        return 1;
    }
    
    std::cout << "Connected to KVDB server" << std::endl;
    
    // 基本操作示例
    std::cout << "\n=== Basic Operations ===" << std::endl;
    
    // PUT操作
    auto put_result = client->put("user:1001", "Alice");
    if (put_result.success) {
        std::cout << "PUT user:1001 = Alice" << std::endl;
    } else {
        std::cerr << "PUT failed: " << put_result.error_message << std::endl;
    }
    
    // GET操作
    auto get_result = client->get("user:1001");
    if (get_result.success) {
        std::cout << "GET user:1001 = " << get_result.value << std::endl;
    } else {
        std::cerr << "GET failed: " << get_result.error_message << std::endl;
    }
    
    // 批量操作示例
    std::cout << "\n=== Batch Operations ===" << std::endl;
    
    std::vector<KeyValue> batch_data = {
        {"user:1002", "Bob"},
        {"user:1003", "Charlie"},
        {"user:1004", "David"}
    };
    
    auto batch_put_result = client->batch_put(batch_data);
    if (batch_put_result.success) {
        std::cout << "Batch PUT completed" << std::endl;
    }
    
    std::vector<std::string> keys = {"user:1001", "user:1002", "user:1003"};
    auto batch_get_result = client->batch_get(keys);
    if (batch_get_result.success) {
        std::cout << "Batch GET results:" << std::endl;
        for (const auto& kv : batch_get_result.value) {
            std::cout << "  " << kv.key << " = " << kv.value << std::endl;
        }
    }
    
    // 扫描操作示例
    std::cout << "\n=== Scan Operations ===" << std::endl;
    
    auto prefix_scan_result = client->prefix_scan("user:", 10);
    if (prefix_scan_result.success) {
        std::cout << "Prefix scan results:" << std::endl;
        for (const auto& kv : prefix_scan_result.value) {
            std::cout << "  " << kv.key << " = " << kv.value << std::endl;
        }
    }
    
    // 快照操作示例
    std::cout << "\n=== Snapshot Operations ===" << std::endl;
    
    auto snapshot_result = client->create_snapshot();
    if (snapshot_result.success) {
        auto snapshot = snapshot_result.value;
        std::cout << "Created snapshot: " << snapshot->id() << std::endl;
        
        // 在快照上读取数据
        auto snapshot_get_result = client->get_at_snapshot("user:1001", snapshot);
        if (snapshot_get_result.success) {
            std::cout << "Snapshot GET user:1001 = " << snapshot_get_result.value << std::endl;
        }
        
        // 释放快照
        client->release_snapshot(snapshot);
        std::cout << "Released snapshot" << std::endl;
    }
    
    // 异步操作示例
    std::cout << "\n=== Async Operations ===" << std::endl;
    
    auto async_put = client->put_async("async:key", "async_value");
    auto async_get = client->get_async("user:1001");
    
    // 等待异步操作完成
    auto async_put_result = async_put.get();
    auto async_get_result = async_get.get();
    
    if (async_put_result.success) {
        std::cout << "Async PUT completed" << std::endl;
    }
    
    if (async_get_result.success) {
        std::cout << "Async GET result: " << async_get_result.value << std::endl;
    }
    
    // 订阅示例
    std::cout << "\n=== Subscription Example ===" << std::endl;
    
    auto subscription_result = client->subscribe("user:*", 
        [](const std::string& key, const std::string& value, const std::string& operation) {
            std::cout << "Subscription event: " << operation << " " << key << " = " << value << std::endl;
        }, true);
    
    if (subscription_result.success) {
        auto subscription = subscription_result.value;
        std::cout << "Started subscription" << std::endl;
        
        // 等待一段时间接收事件
        std::this_thread::sleep_for(std::chrono::seconds(2));
        
        // 触发一些事件
        client->put("user:1005", "Eve");
        client->del("user:1004");
        
        std::this_thread::sleep_for(std::chrono::seconds(1));
        
        // 取消订阅
        subscription->cancel();
        std::cout << "Cancelled subscription" << std::endl;
    }
    
    // 管理操作示例
    std::cout << "\n=== Management Operations ===" << std::endl;
    
    auto stats_result = client->get_stats();
    if (stats_result.success) {
        std::cout << "Database stats:" << std::endl;
        for (const auto& stat : stats_result.value) {
            std::cout << "  " << stat.first << " = " << stat.second << std::endl;
        }
    }
    
    // 断开连接
    client->disconnect();
    std::cout << "\nDisconnected from server" << std::endl;
    
    return 0;
}

File: sdk/java/src/main/java/com/kvdb/client/examples/BasicExample.java
package com.kvdb.client.examples;

import com.kvdb.client.*;
import com.kvdb.client.types.*;

import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

/**
 * KVDB Java客户端基础示例
 */
public class BasicExample {
    
    public static void main(String[] args) {
        // 配置客户端
        ClientConfig config = new ClientConfig()
            .setServerAddress("localhost:50051")
            .setProtocol("grpc")
            .setConnectionTimeout(5000)
            .setRequestTimeout(10000);
        
        // 使用try-with-resources自动关闭客户端
        try (KVDBClient client = new KVDBClient(config)) {
            
            // 连接到服务器
            if (client.connect()) {
                System.out.println("Connected to KVDB server");
            } else {
                System.err.println("Failed to connect to server");
                return;
            }
            
            // 基本操作示例
            basicOperationsExample(client);
            
            // 批量操作示例
            batchOperationsExample(client);
            
            // 扫描操作示例
            scanOperationsExample(client);
            
            // 快照操作示例
            snapshotOperationsExample(client);
            
            // 异步操作示例
            asyncOperationsExample(client);
            
            // 订阅示例
            subscriptionExample(client);
            
            // 管理操作示例
            managementOperationsExample(client);
            
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    private static void basicOperationsExample(KVDBClient client) throws KVDBException {
        System.out.println("\n=== Basic Operations Example ===");
        
        // PUT操作
        client.put("user:1001", "Alice");
        System.out.println("PUT user:1001 = Alice");
        
        // GET操作
        String value = client.get("user:1001");
        System.out.println("GET user:1001 = " + value);
        
        // DELETE操作
        client.delete("user:1001");
        System.out.println("DELETE user:1001");
        
        // 验证删除
        value = client.get("user:1001");
        System.out.println("GET user:1001 after delete = " + value);
    }
    
    private static void batchOperationsExample(KVDBClient client) throws KVDBException {
        System.out.println("\n=== Batch Operations Example ===");
        
        // 批量PUT
        List<KeyValue> pairs = Arrays.asList(
            new KeyValue("user:1001", "Alice"),
            new KeyValue("user:1002", "Bob"),
            new KeyValue("user:1003", "Charlie"),
            new KeyValue("user:1004", "David")
        );
        
        client.batchPut(pairs);
        System.out.println("Batch PUT " + pairs.size() + " pairs");
        
        // 批量GET
        List<String> keys = Arrays.asList("user:1001", "user:1002", "user:1003");
        List<KeyValue> results = client.batchGet(keys);
        
        System.out.println("Batch GET results:");
        for (KeyValue kv : results) {
            System.out.println("  " + kv.getKey() + " = " + kv.getValue());
        }
    }
    
    private static void scanOperationsExample(KVDBClient client) throws KVDBException {
        System.out.println("\n=== Scan Operations Example ===");
        
        // 前缀扫描
        List<KeyValue> results = client.prefixScan("user:", 10);
        System.out.println("Prefix scan results:");
        for (KeyValue kv : results) {
            System.out.println("  " + kv.getKey() + " = " + kv.getValue());
        }
        
        // 范围扫描
        ScanOptions scanOptions = new ScanOptions()
            .setStartKey("user:1000")
            .setEndKey("user:1999")
            .setLimit(5);
        
        results = client.scan(scanOptions);
        System.out.println("Range scan results:");
        for (KeyValue kv : results) {
            System.out.println("  " + kv.getKey() + " = " + kv.getValue());
        }
    }
    
    private static void snapshotOperationsExample(KVDBClient client) throws KVDBException {
        System.out.println("\n=== Snapshot Operations Example ===");
        
        // 创建快照
        Snapshot snapshot = client.createSnapshot();
        System.out.println("Created snapshot: " + snapshot.getSnapshotId());
        
        // 在快照上读取数据
        String value = client.getAtSnapshot("user:1001", snapshot);
        System.out.println("Snapshot GET user:1001 = " + value);
        
        // 修改数据
        client.put("user:1001", "Alice_Modified");
        
        // 当前读取
        String currentValue = client.get("user:1001");
        System.out.println("Current GET user:1001 = " + currentValue);
        
        // 快照读取（应该是旧值）
        String snapshotValue = client.getAtSnapshot("user:1001", snapshot);
        System.out.println("Snapshot GET user:1001 = " + snapshotValue);
        
        // 释放快照
        client.releaseSnapshot(snapshot);
        System.out.println("Released snapshot");
    }
    
    private static void asyncOperationsExample(KVDBClient client) throws Exception {
        System.out.println("\n=== Async Operations Example ===");
        
        // 异步PUT操作
        CompletableFuture<Boolean>[] putFutures = new CompletableFuture[5];
        for (int i = 0; i < 5; i++) {
            putFutures[i] = client.putAsync("async:key" + i, "value" + i);
        }
        
        // 等待所有PUT操作完成
        CompletableFuture.allOf(putFutures).get(10, TimeUnit.SECONDS);
        System.out.println("Completed 5 async PUT operations");
        
        // 异步GET操作
        CompletableFuture<String>[] getFutures = new CompletableFuture[5];
        for (int i = 0; i < 5; i++) {
            getFutures[i] = client.getAsync("async:key" + i);
        }
        
        // 等待所有GET操作完成
        CompletableFuture.allOf(getFutures).get(10, TimeUnit.SECONDS);
        System.out.println("Async GET results:");
        for (int i = 0; i < 5; i++) {
            String value = getFutures[i].get();
            System.out.println("  async:key" + i + " = " + value);
        }
    }
    
    private static void subscriptionExample(KVDBClient client) throws Exception {
        System.out.println("\n=== Subscription Example ===");
        
        // 开始订阅
        int subscriptionId = client.subscribe("user:*", event -> {
            System.out.println("Subscription event: " + event.getOperation() + 
                             " " + event.getKey() + " = " + event.getValue());
        }, true);
        
        System.out.println("Started subscription " + subscriptionId);
        
        // 等待一段时间
        Thread.sleep(1000);
        
        // 触发一些事件
        client.put("user:1005", "Eve");
        client.put("user:1006", "Frank");
        client.delete("user:1004");
        
        // 等待事件处理
        Thread.sleep(2000);
        
        // 取消订阅
        client.cancelSubscription(subscriptionId);
        System.out.println("Cancelled subscription");
    }
    
    private static void managementOperationsExample(KVDBClient client) throws KVDBException {
        System.out.println("\n=== Management Operations Example ===");
        
        // 获取统计信息
        DatabaseStats stats = client.getStats();
        System.out.println("Database stats:");
        System.out.println("  Memtable size: " + stats.getMemtableSize());
        System.out.println("  WAL size: " + stats.getWalSize());
        System.out.println("  Cache hit rate: " + stats.getCacheHitRate());
        System.out.println("  Active snapshots: " + stats.getActiveSnapshots());
        
        // 刷新数据
        client.flush();
        System.out.println("Flushed data to disk");
        
        // 压缩数据
        client.compact();
        System.out.println("Compacted data");
    }
}